// @ngrx-addons/persist-state/esm2022/lib/persist-state.mjs
import { Inject, Injectable } from '@angular/core';
import { InitializationStrategy, isEqual } from '@ngrx-addons/common';
import { Store } from '@ngrx/store';
import { Subject, defaultIfEmpty, distinctUntilChanged, filter, from, map, merge, of, skip, switchMap, takeUntil, tap, } from 'rxjs';
import { rehydrate } from './persist-state.actions';
import { PersistStateRootConfig, PersistStateStrategy, } from './persist-state.config';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "./persist-state.config";
import * as i3 from "@ngrx-addons/common";
const rootState = 'root';
export class PersistState {
    #rootConfig;
    #features = new Map();
    #destroyer = new Subject();
    constructor(store, strategy, rootConfig) {
        this.store = store;
        this.strategy = strategy;
        const { states, storageKeyPrefix, ...restConfig } = rootConfig;
        const keyPrefix = storageKeyPrefix ? `${storageKeyPrefix}-` : '';
        this.#rootConfig = { ...restConfig, storageKeyPrefix: keyPrefix, states };
    }
    addRoot() {
        const merged = this.#rootConfig.states?.map((state) => ({
            ...this.defaultStateConfig(state.key),
            ...state,
            key: state.key,
        })) ?? [];
        this.listenOnStates(merged, rootState).subscribe();
    }
    addFeature(feature) {
        if (this.#features.has(feature.key)) {
            return;
        }
        // Remove in case of re-adding
        this.removeFeature(feature.key);
        this.#features.set(feature.key, true);
        const merged = feature.states.map((state) => ({
            ...this.defaultStateConfig(feature.key),
            ...state,
            key: feature.key,
        }));
        this.listenOnStates(merged, feature.key).subscribe();
    }
    removeFeature(key) {
        this.#destroyer.next(key);
        this.#features.delete(key);
    }
    ngOnDestroy() {
        this.#features.forEach((_, key) => {
            this.removeFeature(key);
        });
        this.#destroyer.next(rootState);
        this.#destroyer.complete();
    }
    defaultStateConfig(key) {
        return {
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            storageKey: `${this.#rootConfig.storageKeyPrefix}${key}@store`,
            source: (state) => state,
            runGuard: () => typeof window !== 'undefined',
            migrations: [],
            skip: 1,
        };
    }
    listenOnStates(states, feature) {
        if (states.length === 0) {
            return of(undefined);
        }
        return merge(...states.map((state) => {
            if (!state.runGuard()) {
                return of(undefined);
            }
            const storage = typeof state.storage === 'function' ? state.storage() : state.storage;
            return merge(
            // Restore state from storage
            this.rehydrateWhen(() => from(storage.getItem(state.storageKey))).pipe(filter((value) => !!value), tap((value) => {
                // Run migrations if defined
                if (state.migrations.length) {
                    value = this.runMigrations(value, state.migrations);
                }
                this.store.dispatch(rehydrate({ features: { [state.key]: value } }));
            }), defaultIfEmpty(undefined)), 
            // Save state to storage
            state
                .source(this.store.pipe(map((storeState) => storeState[state.key])))
                .pipe(distinctUntilChanged(isEqual), skip(state.skip), switchMap((value) => storage.setItem(state.storageKey, value))));
        })).pipe(takeUntil(this.#destroyer.pipe(filter((destroyFeature) => destroyFeature === feature))));
    }
    runMigrations(value, migrations) {
        migrations.forEach((migration) => {
            const version = value[(migration.versionKey ?? 'version')];
            if (migration.version === version) {
                value = migration.migrate(value);
            }
        });
        return value;
    }
    rehydrateWhen(input) {
        return this.strategy.when().pipe(switchMap(() => input()));
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: PersistState, deps: [{ token: i1.Store }, { token: PersistStateStrategy }, { token: i2.PersistStateRootConfig }], target: i0.ɵɵFactoryTarget.Injectable }); }
    /** @nocollapse */ static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: PersistState }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: PersistState, decorators: [{
            type: Injectable
        }], ctorParameters: () => [{ type: i1.Store }, { type: i3.InitializationStrategy, decorators: [{
                    type: Inject,
                    args: [PersistStateStrategy]
                }] }, { type: i2.PersistStateRootConfig }] });
