// @ngrx-addons/persist-state/esm2022/lib/persist-state.feature.mjs
import { Inject, Injectable } from '@angular/core';
import { PersistState } from './persist-state';
import { PERSIST_STATE_FEATURE_CONFIGS } from './persist-state.config';
import * as i0 from "@angular/core";
import * as i1 from "./persist-state";
export class PersistStateFeature {
    constructor(persistState, configs) {
        this.persistState = persistState;
        this.configs = configs;
    }
    addFeatures() {
        this.configs.forEach((config) => {
            this.persistState.addFeature(config);
        });
    }
    ngOnDestroy() {
        this.configs.forEach((config) => {
            this.persistState.removeFeature(config.key);
        });
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: PersistStateFeature, deps: [{ token: i1.PersistState }, { token: PERSIST_STATE_FEATURE_CONFIGS }], target: i0.ɵɵFactoryTarget.Injectable }); }
    /** @nocollapse */ static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: PersistStateFeature }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: PersistStateFeature, decorators: [{
            type: Injectable
        }], ctorParameters: () => [{ type: i1.PersistState }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [PERSIST_STATE_FEATURE_CONFIGS]
                }] }] });
