// @ngrx-addons/persist-state/esm2022/lib/persist-state.provider.mjs
import { ENVIRONMENT_INITIALIZER, inject, makeEnvironmentProviders, } from '@angular/core';
import { BeforeAppInit, afterAppInitProvider } from '@ngrx-addons/common';
import { META_REDUCERS } from '@ngrx/store';
import { PersistState } from './persist-state';
import { FEATURE_PERSIST_STATE, PERSIST_STATE_FEATURE_CONFIGS, PersistStateRootConfig, PersistStateStrategy, ROOT_PERSIST_STORE, } from './persist-state.config';
import { PersistStateFeature } from './persist-state.feature';
import { persistStateReducer } from './persist-state.meta-reducer';
export const _providePersistStore = (config = {}) => {
    return [
        PersistState,
        { provide: PersistStateRootConfig, useValue: config },
        {
            provide: META_REDUCERS,
            useValue: persistStateReducer,
            multi: true,
        },
        afterAppInitProvider,
        {
            provide: PersistStateStrategy,
            useExisting: config.strategy ?? BeforeAppInit,
        },
    ];
};
export const _providePersistState = (config) => {
    return [
        {
            provide: PERSIST_STATE_FEATURE_CONFIGS,
            useValue: config,
            multi: true,
        },
        PersistStateFeature,
    ];
};
/**
 * Persists the global store.
 * These providers cannot be used at the component level.
 *
 * @usageNotes
 *
 * ### Providing the global Persist Store
 *
 * ```ts
 * bootstrapApplication(AppComponent, {
 *   providers: [provideStore(), providePersistStore()],
 * });
 * ```
 */
export const providePersistStore = (config = {}) => {
    return makeEnvironmentProviders([
        ..._providePersistStore(config),
        {
            provide: ROOT_PERSIST_STORE,
            useFactory: () => {
                inject(PersistState).addRoot();
            },
        },
        {
            provide: ENVIRONMENT_INITIALIZER,
            multi: true,
            useFactory: () => () => inject(ROOT_PERSIST_STORE),
        },
    ]);
};
/**
 * Persists additional slices of state.
 * These providers cannot be used at the component level.
 *
 * @usageNotes
 *
 * ### Persist Store Features
 *
 * ```ts
 * const booksRoutes: Route[] = [
 *   {
 *     path: '',
 *     providers: [provideState('books', booksReducer), providePersistState({ key: 'books' })],
 *     children: [
 *       { path: '', component: BookListComponent },
 *       { path: ':id', component: BookDetailsComponent },
 *     ],
 *   },
 * ];
 * ```
 */
export const providePersistState = (config) => {
    return makeEnvironmentProviders([
        ..._providePersistState(config),
        {
            provide: FEATURE_PERSIST_STATE,
            useFactory: () => {
                inject(PersistStateFeature).addFeatures();
            },
        },
        {
            provide: ENVIRONMENT_INITIALIZER,
            multi: true,
            useFactory: () => () => inject(FEATURE_PERSIST_STATE),
        },
    ]);
};
