// @ngrx-addons/persist-state/esm2022/lib/persist-state.module.mjs
import { NgModule } from '@angular/core';
import { PersistStateFeatureModule } from './persist-state.feature.module';
import { _providePersistStore, _providePersistState, } from './persist-state.provider';
import { PersistStateRootModule } from './persist-state.root.module';
import * as i0 from "@angular/core";
export class PersistStateModule {
    static forRoot(config = {}) {
        return {
            ngModule: PersistStateRootModule,
            providers: [..._providePersistStore(config)],
        };
    }
    static forFeature(config) {
        return {
            ngModule: PersistStateFeatureModule,
            providers: [..._providePersistState(config)],
        };
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: PersistStateModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    /** @nocollapse */ static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.0.3", ngImport: i0, type: PersistStateModule }); }
    /** @nocollapse */ static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: PersistStateModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: PersistStateModule, decorators: [{
            type: NgModule
        }] });
