// @ngrx-addons/persist-state/esm2022/lib/storage.mjs
import { of } from 'rxjs';
export const noopStorage = {
    getItem: () => of(null),
    setItem: () => of(true),
    removeItem: () => of(true),
};
export const createStorage = (storage) => {
    if (!storage) {
        return noopStorage;
    }
    return {
        getItem: (key) => {
            const v = storage.getItem(key);
            return of(v ? JSON.parse(v) : v);
        },
        setItem: (key, value) => {
            storage.setItem(key, JSON.stringify(value));
            return of(true);
        },
        // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
        removeItem: (key) => {
            storage.removeItem(key);
            return of(true);
        },
    };
};
// we need to wrap the access to window.localStorage and window.sessionStorage in a try catch
// because localStorage can be disabled, or be denied by a security rule
// as soon as we access the property, it throws an error
const tryGetLocalStorage = () => {
    try {
        if (typeof localStorage !== 'undefined') {
            return localStorage;
        }
    }
    catch {
        // Ignore error
    }
    return undefined;
};
export const localStorageStrategy = createStorage(tryGetLocalStorage());
const tryGetSessionStorage = () => {
    try {
        if (typeof sessionStorage !== 'undefined') {
            return sessionStorage;
        }
    }
    catch {
        // Ignore error
    }
    return undefined;
};
export const sessionStorageStrategy = createStorage(tryGetSessionStorage());
