// @ngrx-addons/common/esm2022/lib/strategies.mjs
import { APP_BOOTSTRAP_LISTENER, inject, Injectable } from '@angular/core';
import { first, of, ReplaySubject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * Interface for strategies implementing way of initialization
 * of the state.
 */
export class InitializationStrategy {
}
/**
 * Strategy initializing state after whole angular application is initialized.
 */
export class AfterAppInit {
    #initialized = new ReplaySubject(1);
    when() {
        return this.#initialized.pipe(first());
    }
    /**
     * Mark strategy as initialized. Meant to be called once whole angular
     * application is initialized.
     */
    markAsInitialized() {
        this.#initialized.next();
    }
    /**
     * Cleanup resources.
     */
    ngOnDestroy() {
        this.#initialized.complete();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: AfterAppInit, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: AfterAppInit, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: AfterAppInit, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
export const afterAppInitProvider = {
    provide: APP_BOOTSTRAP_LISTENER,
    multi: true,
    useFactory: (afterInit = inject(AfterAppInit)) => () => {
        afterInit.markAsInitialized();
    },
};
/**
 * Strategy initializing state as soon as it possible, before angular
 * application is initialized.
 */
export class BeforeAppInit {
    when() {
        return of(void 0);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: BeforeAppInit, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: BeforeAppInit, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: BeforeAppInit, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
